platform/Sqlite.roc

```diff
@@ -0,0 +1,494 @@
+
+    helper! []
+
+decoder : (Value -> Result a (SqlDecodeErr err)) -> (Str -> SqlDecode a err)
+decoder = \fn -> \name ->
+    @SqlDecode \cols ->
+
+        found = List.findFirstIndex cols \x -> x == name
+        when found is
+            Ok index ->
+                \stmt ->
+                    try column_value! stmt index
+                    |> fn
+
+            Err NotFound ->
+                \_ ->
+                    Err (FieldNotFound name)
+
+tagged_value : Str -> SqlDecode Value []
+tagged_value = decoder \val ->
+    Ok val
+
+to_unexpected_type_err = \val ->
+    type =
+        when val is
+            Integer _ -> Integer
+            Real _ -> Real
+            String _ -> String
+            Bytes _ -> Bytes
+            Null -> Null
+    Err (UnexpectedType type)
+
+UnexpectedTypeErr : [UnexpectedType [Integer, Real, String, Bytes, Null]]
+
+str : Str -> SqlDecode Str UnexpectedTypeErr
+str = decoder \val ->
+    when val is
+        String s -> Ok s
+        _ -> to_unexpected_type_err val
+
+bytes : Str -> SqlDecode (List U8) UnexpectedTypeErr
+bytes = decoder \val ->
+    when val is
+        Bytes b -> Ok b
+        _ -> to_unexpected_type_err val
+
+int_decoder : (I64 -> Result a err) -> (Str -> SqlDecode a [FailedToDecodeInteger err]UnexpectedTypeErr)
+int_decoder = \cast ->
+    decoder \val ->
+        when val is
+            Integer i -> cast i |> Result.mapErr FailedToDecodeInteger
+            _ -> to_unexpected_type_err val
+
+real_decoder : (F64 -> Result a err) -> (Str -> SqlDecode a [FailedToDecodeReal err]UnexpectedTypeErr)
+real_decoder = \cast ->
+    decoder \val ->
+        when val is
+            Real r -> cast r |> Result.mapErr FailedToDecodeReal
+            _ -> to_unexpected_type_err val
+
+i64 : Str -> SqlDecode I64 [FailedToDecodeInteger []]UnexpectedTypeErr
+i64 = int_decoder Ok
+
+i32 : Str -> SqlDecode I32 [FailedToDecodeInteger [OutOfBounds]]UnexpectedTypeErr
+i32 = int_decoder Num.toI32Checked
+
+i16 : Str -> SqlDecode I16 [FailedToDecodeInteger [OutOfBounds]]UnexpectedTypeErr
+i16 = int_decoder Num.toI16Checked
+
+i8 : Str -> SqlDecode I8 [FailedToDecodeInteger [OutOfBounds]]UnexpectedTypeErr
+i8 = int_decoder Num.toI8Checked
+
+u64 : Str -> SqlDecode U64 [FailedToDecodeInteger [OutOfBounds]]UnexpectedTypeErr
+u64 = int_decoder Num.toU64Checked
+
+u32 : Str -> SqlDecode U32 [FailedToDecodeInteger [OutOfBounds]]UnexpectedTypeErr
+u32 = int_decoder Num.toU32Checked
+
+u16 : Str -> SqlDecode U16 [FailedToDecodeInteger [OutOfBounds]]UnexpectedTypeErr
+u16 = int_decoder Num.toU16Checked
+
+u8 : Str -> SqlDecode U8 [FailedToDecodeInteger [OutOfBounds]]UnexpectedTypeErr
+u8 = int_decoder Num.toU8Checked
+
+f64 : Str -> SqlDecode F64 [FailedToDecodeReal []]UnexpectedTypeErr
+f64 = real_decoder Ok
+
+f32 : Str -> SqlDecode F32 [FailedToDecodeReal []]UnexpectedTypeErr
+f32 = real_decoder (\x -> Num.toF32 x |> Ok)
+
+# TODO: Mising Num.toDec and Num.toDecChecked
+# dec = realSqlDecoder Ok
```

Comment:
Can we use [`Std.toDec : Str -> Result Dec [InvalidNumStr]`](https://www.roc-lang.org/builtins/Str#toDec)

Reply:
I guess we could. Float -> Str -> Dec. That said, we probably should just add `Num.toDec`. for floats, it just means multiplying by `10^18` then casting to an `i128`.

Reply:
We already do this in the zig bitcode. Just isn't wired up/exposed.